composer/src/composer.rs

```diff
@@ -1368,16 +1406,16 @@ impl ComposeTest {
             EndpointSettings {
                 network_id: Some(self.network_id.to_string()),
                 ipam_config: Some(EndpointIpamConfig {
-                    ipv4_address: Some(ipv4.to_string()),
+                    ipv4_address: ipv4.map(|ip| ip.to_string()),
                     ..Default::default()
                 }),
                 aliases: spec.network_aliases.clone(),
                 ..Default::default()
             },
         );
 
-        let mut env = self.spec_environment(spec);
-        env.push(format!("MY_POD_IP={ipv4}"));
+        let mut env = spec.environment();
+        env.push(format!("MY_POD_IP={:?}", ipv4.unwrap_or(Ipv4Addr::new(127, 0, 0, 1))));
```

Comment:
Ah so every container gets the same ip?

Reply:
> hmm can we leave the containers as ip and then just unwrap to 127.0.0.1 here itself to avoid changing the public api?

Are you suggesting:
1. set the container's IP from here in hashmap of (containerId, ip address) as 127.0.0.1, and
2. let the `grpc_handle()` API internals be as it is today and expect the ip available always while returning SocketAddr?

```
format!("{}:10124", v.1 .1.unwrap_or(Ipv4Addr::new(127, 0, 0, 1)))
                        .parse::<std::net::SocketAddr>()
                        .unwrap()
```

Reply:
Yeah, basically just make sure the ip is set on the container hashmap

Reply:
> Yeah, basically just make sure the ip is set on the container hashmap

I think I had tried that and ran into some issue. Since I don't remember what trouble I faced, I'll try once again.

Reply:
I've simplified now, keeping hashmap as-is. Setting ip as localhost early on, but handling based on network mode just before calling docker create_container.